include/deal.II/base/mpi_compute_index_owner_internal.h

```diff
@@ -0,0 +1,799 @@
+           */
+          const MPI_Comm &comm;
+
+          /**
+           * The present MPI rank.
+           */
+          const unsigned int my_rank;
+
+          /**
+           * The total number of ranks participating in the MPI communicator
+           * `comm`.
+           */
+          const unsigned int n_procs;
+
+          /**
+           * Controls whether the origin of ghost owner should also be
+           * stored. If true, it will be added into `requesters` and can be
+           * queried by `get_requesters()`.
+           */
+          const bool track_index_requests;
+
+          /**
+           * The result of the index owner computation: To each index
+           * contained in `indices_to_look_up`, this vector contains the MPI
+           * rank of the owner in `owned_indices`.
+           */
+          std::vector<unsigned int> &owning_ranks;
+
+          /**
+           * Keeps track of the origin of the requests. The layout of the data
+           * structure is as follows: The outermost vector has as many entries
+           * as Dictionary::actually_owning_rank_list and represents the
+           * information we should send back to the owners from the present
+           * dictionary entry. The second vector then collects a list of MPI
+           * ranks that have requested data, using the rank in the first pair
+           * entry and a list of index ranges as the second entry.
+           */
+          std::vector<std::vector<
+            std::pair<unsigned int,
+                      std::vector<std::pair<unsigned int, unsigned int>>>>>
+            requesters;
+
+          /**
+           * The dictionary handling the requests.
+           */
+          Dictionary dict;
+
+          /**
+           * Temporary data for the computation of the owner.
+           */
+          std::map<unsigned int, std::vector<types::global_dof_index>> temp;
+
+          /**
+           * The field where the indices for incoming data from the process
+           * are stored.
+           */
+          std::map<unsigned int, std::vector<unsigned int>> recv_indices;
+
+          /**
+           * Implementation of
+           * Utilities::MPI::ConsensusAlgorithmProcess::process_request(),
+           * adding the owner of a particular index in request_buffer (and
+           * keeping track of who requested a particular index in case that
+           * information is also desired).
+           */
+          virtual void
+          process_request(
+            const unsigned int                          other_rank,
+            const std::vector<types::global_dof_index> &buffer_recv,
+            std::vector<unsigned int> &                 request_buffer) override
+          {
+            Assert(buffer_recv.size() % 2 == 0, ExcInternalError());
+            unsigned int owner_index = 0;
+            for (unsigned int j = 0; j < buffer_recv.size(); j += 2)
+              for (auto i = buffer_recv[j]; i < buffer_recv[j + 1]; i++)
```

Comment:
``buffer_recv`` conceptually is a vector of ``pair<dof_index,dof_index>``, is that possible to make that change?

Reply:
Not easily at least: The problem is that we use `mpi_type_id(T1)` in `ConsensusAlgorithm` for sending the MPI, and we do not have that for `pair<dof_index,dof_index>`. So we would either have to define an MPI type for a pair or use `MPI_BYTE` as a type times `sizeof(T1)`. I will temporarily go with the `sizeof(T1)` variant but if @peterrum strongly objects how I misuse the class we should think about an alternative.

Reply:
@tjhei It used to be a pair. I guess if you have look at the first commit regarding this matter you will see that. I moved from pair to an int because of the reasons @kronbichler wrote (when I generalized the class and introduced the template parameter). `sizeof` is somewhat an ugly approach, but I guess it will do the job ;)

Reply:
I have that variant implemented. Let me run some local matrix-free/multigrid tests where this is stressed a lot before I force-push the changes.